File: src/types/grep.ts
import { z } from "zod";

/**
 * Grep — match-centric full-corpus retrieval (ADR-0035).
 *
 * Complements the 8 per-entity `search_*` tools (entity-centric discovery):
 * `grep` answers "where does this text occur?" across ALL author-written
 * text in a World — including the four CharacterPhase fields, which the
 * existing `search_characters` does not cover. Results are occurrence
 * evidence (match counts + before/match/after snippets) grouped by
 * (entityType, entityId, fieldName), not entity summaries.
 *
 * Computed on demand by the `grep` Rust aggregation command (SQL `LIKE`
 * prefilter + in-memory Rust scan with identical ASCII case folding, so
 * the two passes have zero false negatives between them). Never persisted.
 */

// ─── Entity type ──────────────────────────────────────────────────────────

/**
 * The entity types `grep` can sweep: the 8 `search_*` entity types plus
 * `phase` — CharacterPhase is an independent table whose text fields
 * (`name`, `appearance`, `description`, `conversation_style`) are part of
 * the corpus but invisible to `search_characters`. Also the fixed
 * tie-break order for equally-ranked match groups (ADR-0035 §5).
 */
export const grepEntityTypeSchema = z.enum([
  "character",
  "phase",
  "location",
  "item",
  "lore",
  "event",
  "novel",
  "chapter",
  "scene",
]);

export type GrepEntityType = z.infer<typeof grepEntityTypeSchema>;

// ─── Snippet ──────────────────────────────────────────────────────────────

/**
 * One match occurrence rendered as three parts: the matched substring with
 * up to ~40 chars of context on each side. Character-based truncation
 * (UTF-8 boundary safe on the Rust side), no `...` / `【】` marker glyphs —
 * markers collide with characters that occur in prose and mislead the
 * model (ADR-0035 §4).
 */
export const grepSnippetSchema = z.object({
  /** Up to ~40 chars immediately before the match (`""` at text start). */
  before: z.string(),
  /** The matched substring, exactly as it occurs in the text. */
  match: z.string(),
  /** Up to ~40 chars immediately after the match (`""` at text end). */
  after: z.string(),
});

export type GrepSnippet = z.infer<typeof grepSnippetSchema>;

// ─── Match group ──────────────────────────────────────────────────────────

/**
 * All matches of the query within one `(entityType, entityId, fieldName)`
 * field, carrying redundant entity identity so the model can act without
 * a follow-up `get_*` call. `entityId` is always directly usable by the
 * `get_*` family — phase hits are `entityType: "phase"` with
 * `entityId = phaseId`, never an overloaded character encoding
 * (ADR-0035 §3).
 *
 * `characterId` / `characterName` are non-null only when
 * `entityType === "phase"`; they arrive as `null` (not `undefined`) from
 * serde.
 */
export const grepMatchGroupSchema = z.object({
  /** Which entity type the matched field belongs to. */
  entityType: grepEntityTypeSchema,
  /** UUID of the entity (the phase id when `entityType === "phase"`). Polymorphic — not a branded entity id. */
  entityId: z.string(),
  /** Display name of the entity (character name, phase name, scene title, …). */
  entityTitle: z.string(),
  /** Owning character id — non-null only when `entityType === "phase"`. */
  characterId: z.string().nullable(),
  /** Owning character name — non-null only when `entityType === "phase"`. */
  characterName: z.string().nullable(),
  /** Field name the matches were found in (e.g. `"description"`, `"content"`). */
  fieldName: z.string(),
  /** Total matches in this field — not capped, unlike `snippets`. */
  matchCount: z.number().int(),
  /** Up to 3 snippet samples; need more context → fetch the entity via `get_*`. */
  snippets: z.array(grepSnippetSchema).max(3),
});

export type GrepMatchGroup = z.infer<typeof grepMatchGroupSchema>;

// ─── Response ─────────────────────────────────────────────────────────────

/**
 * Result of a `grep` query — one PAGE of a deterministically ordered result
 * set. Groups are sorted by `matchCount` descending, ties broken by fixed
 * entity-type enum order then title ascending — deterministic, so the model
 * can diff two greps across turns AND paginate stably: 50 groups per page,
 * walk `offset` (0, 50, 100, …) while `truncated` is `true`.
 */
export const grepResultSchema = z.object({
  /** Echo of the query string. */
  query: z.string(),
  /** One page of match groups, `matchCount` desc, up to 50 entries. */
  groups: z.array(grepMatchGroupSchema),
  /**
   * FULL group count before pagination — may exceed `groups.length` when
   * further pages exist (pairs with `truncated` to tell the model how much
   * was cut).
   */
  groupCount: z.number().int(),
  /**
   * `true` when more groups exist BEYOND this page
   * (`offset + groups.length < groupCount`) — fetch the next page by
   * passing an increased `offset`.
   */
  truncated: z.boolean(),
});

export type GrepResult = z.infer<typeof grepResultSchema>;

File: src-tauri/src/models/grep.rs
use serde::Serialize;

/// Top-level response of the `grep` command (ADR-0035) — a match-centric,
/// field-grouped sweep across all author-written text of one World.
///
/// Carries the query back verbatim so the response is self-describing in
/// tool transcripts. Note the query is IPC payload only — it is never
/// logged (ADR-0016 never-log applies to query text at any level).
#[derive(Debug, Clone, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct GrepResult {
    /// Echoes the caller's query verbatim.
    pub query: String,
    /// Match groups, sorted by `match_count` desc (ties → fixed entity-type
    /// order, then `entity_title` asc), capped at 50 entries.
    pub groups: Vec<GrepMatchGroup>,
    /// Full group count BEFORE the 50-group cap was applied.
    pub group_count: i64,
    /// `true` when `group_count > groups.len()` — the model should re-query
    /// with a narrower `entityTypes` or a more specific query.
    pub truncated: bool,
}

/// One match group: every occurrence of the query within ONE field of ONE
/// entity, collapsed to a count plus up to 3 snippets (ADR-0035 §3 — a
/// protagonist name appearing 50× in one scene stays one group, not 50
/// occurrence records).
#[derive(Debug, Clone, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct GrepMatchGroup {
    /// One of the 9 grep corpus types (`"character"`, `"phase"`, …).
    pub entity_type: String,
    /// Directly usable by the `get_*` family. For phase groups this is the
    /// PHASE id — never an overloaded `character` + non-null phase id
    /// encoding (ADR-0035 §3).
    pub entity_id: String,
    /// The entity's `name` (or `title` for novel/chapter/scene); for phase
    /// groups, the phase's own name.
    pub entity_title: String,
    /// PHASE groups only: the owning character's id. `None` otherwise.
    pub character_id: Option<String>,
    /// PHASE groups only: the owning character's name. `None` otherwise.
    pub character_name: Option<String>,
    /// The field the matches live in (e.g. `"content"`, `"aliases"`, `"tags"`).
    pub field_name: String,
    /// Total non-overlapping occurrences (plain-text fields) or matching
    /// elements (JSON-array fields) — uncapped; preserves the "mentioned
    /// 50 times" signal beyond the snippet sample.
    pub match_count: i64,
    /// The first `SNIPPETS_PER_GROUP` occurrences as context snippets.
    pub snippets: Vec<GrepSnippet>,
}

/// Three-part context snippet around one occurrence (ADR-0035 §4): up to 40
/// CHARS per side, no `...`/`【】` marker glyphs — markers collide with
/// characters that occur in prose and mislead the model.
#[derive(Debug, Clone, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct GrepSnippet {
    /// Up to 40 chars immediately before the match (`""` when none).
    pub before: String,
    /// The matched text, original case. `match` is a Rust keyword, hence
    /// the raw identifier — serde still emits the field as `"match"`.
    pub r#match: String,
    /// Up to 40 chars immediately after the match (`""` when none).
    pub after: String,
}
